Wait for the new child to be ready.

iwanbk opened the pull request:
Wait for the new child notification about it's readyness
before killing the old one.

Child send the notifcation to socketmaster using SIGUSR1 signal.

**Why we need this feature**

`socketmaster` currently wait for `-start` milliseconds before killing the old process.
It doesn't care whether the new process is ready or not.

It created downtime when the new process took a long time to initialise.

One of the solution is to set `-start` argument to a very long time. But i don't think that it is the good solution with these reasons:
- `race condition` like this should not be fixed with timing, it should be fixed with coordination or lock in case of single process
- what if we want the deployment to be finished as soon as possible
- what if for some reason, the initialise still take longer time
- it simply not elegant

**use case**

`feeds` got this issue, i also got info that some other teams also face same issue.

**added description on readme**
## Wait Child Notif

When activated using `-wait-child-notif` flag, `socketmaster` will wait for the newly created process
to send `SIGUSR1` before killing the old one.

Note that `-wait-child-notif` still respect `-start` argument.

In order to use this feature, the app server need to do these:
- ~add support for `socketmaster-pid` command line argument
  `socketmaster` will start the app server with `socketmaster-id {PID_OF_SOCKETMASTER}`
  argument~
- read `SOCKETMASTER_PID` environment variable to get `socketmaster` process ID
- Send `SIGUSR1` signal after the appserver is ready

*Example in Go*

~add support for `socketmaster-pid` command line argument~

Send `SIGUSR1` signal after the appserver is ready
```go
import (
  "github.com/tokopedia/socketmaster/child"
)
go child.NotifyMaster()
```

albertwidi commented:
I do agree that the current version is problematic and can go wrong in many ways, the creator is also working on the new things since 4 years ago https://news.ycombinator.com/item?id=8772636. This is actually a hack for reloading the binaries. 

Worst, any packets that are not forwarded via the socketmaster fds(3) is dropped. For example if a server has `http` and `grpc` server. Usually we use the http address and port for socketmaster, and all packets into grpc are dropped.

The PR is LGTM, do you have any shell script to test the feature?

iwanbk commented:
> The PR is LGTM, do you have any shell script to test the feature?

Sure, we need 3 terminals for this test 

(1) I put below script in `/Users/nakama/tmp/slowserver/` directory
```go
package main

import (
	"context"
	"fmt"
	"log"
	"net"
	"net/http"
	"os"
	"os/signal"
	"syscall"
	"time"

	"github.com/tokopedia/socketmaster/child"
)

func main() {
	pid := os.Getpid()

	log.Printf("[app:%v] initializing", pid)

	handler := func(w http.ResponseWriter, r *http.Request) {
		log.Printf("[app:%v] got request\n", pid)
		fmt.Fprintf(w, "Hello, you've requested: %s\n", r.URL.Path)
	}

	time.Sleep(20 * time.Second) // simulate slow initialization
	log.Printf("[app:%v] start listening", pid)
	server := http.Server{
		Handler: http.HandlerFunc(handler),
	}
	lis := getListener()

	go func() {
		// notify master
		err := child.NotifyMaster()
		if err != nil {
			log.Printf("failed to notify master:%v")
		}

		// wait for the SIGTERM
		sigs := make(chan os.Signal, 1)
		signal.Notify(sigs, syscall.SIGTERM)
		<-sigs
		log.Printf("[app:%v] got SIGTERM. Shut me down....\n", pid)
		server.Shutdown(context.Background())
	}()
	server.Serve(lis)
}

func getListener() net.Listener {
	sockFD := 3
	file := os.NewFile(uintptr(sockFD), "listener")
	fl, err := net.FileListener(file)
	if err != nil {
		log.Fatalf("failed to create listener:%v", err)
	}
	return fl
}

```
build
```bash
go build
```

after we start socketmaster, we execute `curl` to test it
```bash
curl -m 5  http://localhost:3030/
```
it will timeout in 5 seconds

(2) use `socketmaster` to start the process
```bash
$ ./socketmaster --command=/Users/nakama/tmp/slowserver/slowserver -listen tcp://:3030 -wait-child-notif=true
```

(3) restart the socketmaster
```bash
$ ps aux | grep socketmaster
nakama           83923   0.0  0.0  4307308      8 s024  S+    2:31PM   0:01.36 vim socketmaster.go
nakama            2721   0.0  0.0  4258468    192 s021  R+   11:20AM   0:00.00 grep socketmaster
nakama            2717   0.0  0.0  4377636   2768 s017  S+   11:19AM   0:00.01 ./socketmaster --command=/Users/nakama/tmp/slowserver/slowserver -listen tcp://:3030 -wait-child-notif=false                
nakama:~/pathgo/src/github.com/tokopedia/socketmaster
$ kill -SIGHUP 2717
```

then we can compare it with starting `socketmaster` without this `-wait-child-notif=true` option

iwanbk commented:
> I do agree that the current version is problematic and can go wrong in many ways, the creator is also working on the new things since 4 years ago https://news.ycombinator.com/item?id=8772636. 

I'll surely check it ASAP, thanks for the info

> For example if a server has http and grpc server. Usually we use the http address and port for socketmaster, and all packets into grpc are dropped.

Yep, i noticed it yesterday when i was checking `socketmaster` yesterday

awmanoj commented:
Looks good.

iwanbk commented:
@albertwidi 
I've just updated this PR at 7be0a2eebb89a19b11a7bef94662f5bf6a87f472
Now it uses environment variable to tell client about socketmaster pid.
Much simpler than using command line argument.

I've updated my sampe script as well

albertwidi commented:
Cool, do you need this ASAP(or you already build the binary for your own team)? I have some free time in the evening and can help to test more.

iwanbk commented:
No worries, please take your time :)

iwanbk commented:
> Usually we use the http address and port for socketmaster, and all packets into grpc are dropped.

Curious question, why we do this? (not run grpc behind `socketmaster`)

bmishra commented:
we use the same binary to launch http and grpc services, if we build and run the two services separately we can use socketmaster

iwanbk commented:
> we use the same binary to launch http and grpc services, if we build and run the two services separately we can use socketmaster

yes, my team also separate it.
